libcontainer/init_linux.go

```diff
@@ -79,47 +79,77 @@ func newContainerInit(t initType, pipe *os.File, consoleSocket *os.File, fifoFd
 	if err := json.NewDecoder(pipe).Decode(&config); err != nil {
 		return nil, err
 	}
-	if err := populateProcessEnvironment(config.Env); err != nil {
+	environmentMap, err := populateEnvironmentMap(config.Env)
+	if err != nil {
 		return nil, err
 	}
+
 	switch t {
 	case initSetns:
 		return &linuxSetnsInit{
-			pipe:          pipe,
-			consoleSocket: consoleSocket,
-			config:        config,
+			pipe:           pipe,
+			consoleSocket:  consoleSocket,
+			config:         config,
+			environmentMap: environmentMap,
 		}, nil
 	case initStandard:
 		return &linuxStandardInit{
-			pipe:          pipe,
-			consoleSocket: consoleSocket,
-			parentPid:     unix.Getppid(),
-			config:        config,
-			fifoFd:        fifoFd,
+			pipe:           pipe,
+			consoleSocket:  consoleSocket,
+			parentPid:      unix.Getppid(),
+			config:         config,
+			fifoFd:         fifoFd,
+			environmentMap: environmentMap,
 		}, nil
 	}
 	return nil, fmt.Errorf("unknown init type %q", t)
 }
 
-// populateProcessEnvironment loads the provided environment variables into the
-// current processes's environment.
-func populateProcessEnvironment(env []string) error {
-	for _, pair := range env {
-		p := strings.SplitN(pair, "=", 2)
-		if len(p) < 2 {
-			return fmt.Errorf("invalid environment '%v'", pair)
+// parseEnvironmentVariable takes a environment variable with the format
+// key=value, and returns the key and value seperated.
+func parseEnvironmentVariable(env string) (key, value string, err error) {
+	p := strings.SplitN(env, "=", 2)
+	if len(p) < 2 {
+		err = fmt.Errorf("invalid environment '%+v'", env)
+		return
+	}
+	key = p[0]
+	value = p[1]
+	return
+}
+
+// createEnvironmentMap loads the provided environment variables into a
+// map.
+func populateEnvironmentMap(env []string) (map[string]string, error) {
+	environ := os.Environ()
+	mappedEnv := make(map[string]string, len(environ)+len(env))
+
+	for _, pair := range environ {
+		key, _, err := parseEnvironmentVariable(pair)
+		if err != nil {
+			return nil, err
 		}
-		if err := os.Setenv(p[0], p[1]); err != nil {
-			return err
+		mappedEnv[key] = pair
+	}
+	for _, pair := range env {
+		key, _, err := parseEnvironmentVariable(pair)
+		if err != nil {
+			return nil, err
 		}
+		mappedEnv[key] = pair
 	}
-	return nil
+
+	// Certain functions that are used later, such as exec.LookPath, require that the PATH
+	// environment variable is setup with the container's PATH.
+	os.Setenv("PATH", mappedEnv["PATH"])
```

Comment:
Could you be more specific? Wondering it may cause security issues.

Reply:
1. This does not causes any security issues because this is what was done before the patch, too.
2. This appears to be an error because `mappedEnv["PATH"]` contains, for example, `PATH=/usr/bin:/usr/sbin` (rather than `/usr/bin:/usr/sbin`. As a result, the first PATH element will be wrong.